src/backend/distributed/executor/executor.c

```diff
@@ -1343,6 +1375,78 @@ ManageWorkerPool(WorkerPool *workerPool)
 }
 
 
+/*
+ * CheckConnectionTimeout makes sure that the execution enforces the connection
+ * establishment timeout defined by the user (NodeConnectionTimeout).
+ *
+ * The rule is that if a worker pool has already initiated connection establishment
+ * and has not succeeded to finish establishments that are necessary to execute tasks,
+ * take an action. For the types of actions, see the comments in the function.
+ *
+ * Enforcing the timeout per pool (over per session) helps the execution to continue
+ * even if we can establish a single connection as we expect to have target pool size
+ * number of connections. In the end, the executor is capable of using one connection
+ * to execute multiple tasks.
+ */
+static void
+CheckConnectionTimeout(WorkerPool *workerPool)
+{
+	DistributedExecution *execution = workerPool->distributedExecution;
+	TimestampTz poolStartTime = workerPool->poolStartTime;
+	TimestampTz now = GetCurrentTimestamp();
+
+	int initiatedConnectionCount = list_length(workerPool->sessionList);
+	int activeConnectionCount = workerPool->activeConnectionCount;
+	int requiredActiveConnectionCount = 1;
+
+	if (initiatedConnectionCount == 0)
+	{
+		/* no connection has been planned for the pool yet */
+		Assert(poolStartTime == 0);
+		return;
+	}
+
+	/*
+	 * This is a special case where we assign tasks to sessions even before
+	 * the connections are established. So, make sure to apply similar
+	 * restrictions. In this case, make sure that we get all the connections
+	 * established.
+	 */
+	if (ForceMaxQueryParallelization)
+	{
+		requiredActiveConnectionCount = initiatedConnectionCount;
```

Comment:
Is this affected by #2764?

Reply:
(I think not)

Reply:
As far as I can see, no. Because, even with #2764 we're still initiating all the connections at the same time, so it is either `initiatedConnectionCount == 0` or `requiredActiveConnectionCount = initiatedConnectionCount;`.

But, it would be effected if decide to open connections gradually, as you suggested such as with `10ms` wait loop. But, still, for  `ForceMaxQueryParallelization ` we probably wouldn't implement such thing and this should be fine.